### PR TITLE
[Relay][External Codegen] Support data types for CSourceModuleCodegen args and output

### PR DESCRIPTION
**Pull request by soiferj**
Adds a struct, `Output`, to hold information about outputs for `CSourceModuleCodegen`. The codegen then uses this information to generate code with the correct datatype and copy it to the output buffer (if necessary).

@comaniac @zhiics would you mind taking a look?

I added a unit test for dtype, but did not add a unit test for ignoring the `memcpy`. I was having a hard time making the codegen code look clean.

### CODE REVIEW

File: src/relay/backend/contrib/codegen_c/codegen.cc [hunk 1]
```diff
@@ -70,6 +72,12 @@ class CodegenC : public ExprVisitor, public CodegenCBase {
     for (size_t i = 0; i < in_shape.size(); ++i) {
       macro_stream << ", " << in_shape[i];
     }
+
+    auto type_node = call->checked_type().as<TensorTypeNode>();
```
Comment (zhiics) [inline src/relay/backend/contrib/codegen_c/codegen.cc]: `const auto* type_node`

File: src/relay/backend/contrib/codegen_c/codegen_c.h [hunk 2]
```diff
@@ -116,29 +123,30 @@ class CodegenCBase {
    *
    * \endcode
    */
-  void GenerateBackendCFunc(const std::string& func_name, int arg_cnt) {
+  void GenerateBackendCFunc(const std::string& func_name, Array<Var> args, const Output& out) {
```
Comment (zhiics) [inline src/relay/backend/contrib/codegen_c/codegen_c.h]: const Array<Var>& args

File: src/relay/backend/contrib/codegen_c/codegen_c.h [hunk 3]
```diff
@@ -207,17 +215,20 @@ class CodegenCBase {
    *
    * \return The emitted code string.
    */
-  std::string JitImpl(std::string ext_func_id, std::vector<std::string> args,
+  std::string JitImpl(std::string ext_func_id, Array<Var> args,
```
Comment (zhiics) [inline src/relay/backend/contrib/codegen_c/codegen_c.h]: please also use const reference here for all parameters.

File: src/relay/backend/contrib/codegen_c/codegen_c.h [hunk 4]
```diff
@@ -232,24 +243,60 @@ class CodegenCBase {
...
   }
 
+  /*!
+   * \brief Returns dtype string
+   *
+   * \param var Var to get the dtype of
+   *
+   * \return The dtype string.
+   */
+  std::string GetDtypeString(Var var) {
```
Comment (zhiics) [inline src/relay/backend/contrib/codegen_c/codegen_c.h]: const Var&

File: src/relay/backend/contrib/codegen_c/codegen.cc (hunk 1 above)
Comment (soiferj) [reply src/relay/backend/contrib/codegen_c/codegen.cc]: Updated

File: src/relay/backend/contrib/codegen_c/codegen_c.h (hunk 2 above)
Comment (soiferj) [reply src/relay/backend/contrib/codegen_c/codegen_c.h]: Updated

File: src/relay/backend/contrib/codegen_c/codegen_c.h (hunk 3 above)
Comment (soiferj) [reply src/relay/backend/contrib/codegen_c/codegen_c.h]: Updated

File: src/relay/backend/contrib/codegen_c/codegen_c.h (hunk 4 above)
Comment (soiferj) [reply src/relay/backend/contrib/codegen_c/codegen_c.h]: Updated

**Pull request merged**